### PR TITLE
i#1569 AArch64: Fix macros to create AND,ANDs with imms.

### PR DESCRIPTION
**Pull request by fhahn**
Instructions that take logical immediates are not encoded with shifts and the 
macros were broken for them.

Issue: #1569

### CODE REVIEW

File: suite/tests/api/ir_aarch64.c [hunk 1]
```diff
@@ -282,11 +282,33 @@ test_ldar(void *dc)
...
 
 static void
-test_fmov_general(void *dc)
+test_instrs_with_logic_imm(void *dc)
 {
     byte *pc;
     instr_t *instr;
 
+    instr = INSTR_CREATE_and(dc, opnd_create_reg(DR_REG_X10),
+                             opnd_create_reg(DR_REG_X9), OPND_CREATE_INT(0xFFFF));
```
Comment (derekbruening) [inline suite/tests/api/ir_aarch64.c]: (There's an issue on auto-generating the INSTR_CREATE_ macros -- and presumably the tests as well.)

File: core/arch/aarch64/instr_create.h [hunk 2]
```diff
@@ -444,14 +444,34 @@
...
+ * \param dc   The void * dcontext used to allocate memory for the instr_t.
+ * \param rd   The output register.
+ * \param rn   The first input register.
+ * \param rm_or_imm   The second input register or immediate.
+ */
 #define INSTR_CREATE_and(dc, rd, rn, rm_or_imm) \
-  INSTR_CREATE_and_shift(dc, rd, rn, rm_or_imm, OPND_CREATE_LSL(), OPND_CREATE_INT(0))
+  opnd_is_immed(rm_or_imm) ? instr_create_1dst_2src((dc), OP_and, (rd), (rn), (rm_or_imm)) :  \
+                             INSTR_CREATE_and_shift(dc, rd, rn, rm_or_imm, OPND_CREATE_LSL(), \
+                                                    OPND_CREATE_INT(0))
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: Put () around the whole thing for safety

File: core/arch/aarch64/instr_create.h [hunk 3]
```diff
@@ -444,14 +444,34 @@
...
+ * \param dc   The void * dcontext used to allocate memory for the instr_t.
+ * \param rd   The output register.
+ * \param rn   The first input register.
+ * \param rm_or_imm   The second input register or immediate.
+ */
 #define INSTR_CREATE_ands(dc, rd, rn, rm_or_imm) \
-  INSTR_CREATE_ands_shift(dc, rd, rn, rm_or_imm, OPND_CREATE_LSL(), OPND_CREATE_INT(0))
+  opnd_is_immed(rm_or_imm) ? instr_create_1dst_2src((dc), OP_ands, (rd), (rn), (rm_or_imm)) :  \
+                             INSTR_CREATE_ands_shift(dc, rd, rn, rm_or_imm, OPND_CREATE_LSL(), \
+                                                    OPND_CREATE_INT(0))
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: Put () around the whole thing for safety

Comment (fhahn) [reply core/arch/aarch64/instr_create.h]: Done

File: core/arch/aarch64/instr_create.h (hunk 2 above)
Comment (fhahn) [reply core/arch/aarch64/instr_create.h]: Done, thank you very much!

File: suite/tests/api/ir_aarch64.c (hunk 1 above)
Comment (fhahn) [reply suite/tests/api/ir_aarch64.c]: (y) I'll check with @AssadHashmi, not sure how far he got so far. At least for the remaining NEON instructions, I try to generate most things. Except for some instructions  with uncommon operands, like fmov.

**Pull request merged**